lib/ttfunk/table/cff/dict.rb

```diff
@@ -0,0 +1,225 @@
+
+        def encode_operand(operand)
+          bytes = case operand
+                  when Integer
+                    encode_integer(operand)
+                  when Float, BigDecimal
+                    encode_float(operand)
+                  when Sci
+                    encode_sci(operand)
+                  end
+
+          bytes.pack('C*')
+        end
+
+        def encode_integer(int)
+          case int
+          when -107..107
+            [int + 139]
+
+          when 108..1131
+            int -= 108
+            [(int >> 8) + 247, int & 0xFF]
+
+          when -1131..-108
+            int = -int - 108
+            [(int >> 8) + 251, int & 0xFF]
+
+          when -32_768..32_767
+            [28, (int >> 8) & 0xFF, int & 0xFF]
+
+          else
+            encode_integer32(int)
+          end
+        end
+
+        def encode_integer32(int)
+          [
+            29,
+            (int >> 24) & 0xFF,
+            (int >> 16) & 0xFF,
+            (int >> 8) & 0xFF,
+            int & 0xFF
+          ]
+        end
+
+        def encode_float(float)
+          pack_decimal_nibbles(encode_significand(float))
+        end
+
+        def encode_sci(sci)
+          sig_bytes = encode_significand(sci.significand)
+          exp_bytes = encode_exponent(sci.exponent)
+          pack_decimal_nibbles(sig_bytes + exp_bytes)
+        end
+
+        def encode_exponent(exp)
+          return [] if exp == 0
+          [exp > 0 ? 0xB : 0xC, *encode_significand(exp)]
+        end
+
+        def encode_significand(sig)
+          sig.to_s.each_char.with_object([]) do |char, ret|
+            case char
+            when '0'..'9'
+              ret << char.to_i
+            when '.'
+              ret << 0xA
+            when '-'
+              ret << 0xE
+            else
+              break ret
+            end
+          end
+        end
+
+        def pack_decimal_nibbles(nibbles)
+          packed = [30]
+
+          nibbles.each_slice(2).each do |(high_nb, low_nb)|
+            # low_nb can be nil if nibbles contains an odd number of elements
+            low_nb ||= 0xF
+            packed << (high_nb << 4 | low_nb)
+          end
+
+          packed << 0xFF if nibbles.size.even?
+          packed
+        end
+
+        def parse!
+          @dict = {}
+          operands = []
+
+          # @length must be set via the constructor
+          while io.pos < table_offset + length
+            case b_zero = read(1, 'C').first
+            when 12
+              operator = decode_two_byte_operator
+              @dict[operator] = operands
+              operands = []
+            when 0..21
+              @dict[b_zero] = operands
+              operands = []
+            when 28..30, 32..254
+              operands << decode_operand(b_zero)
+            else
+              raise "dict byte value #{b_zero} is reserved"
+            end
+          end
+        end
+
+        def decode_two_byte_operator
+          1200 + read(1, 'C').first
```

Comment:
This is an arbitrary magic number. It doesn't really represent any special or useful value. Maybe it'd be easier to keep operators arrays of numbers? Like, `[12, 3]`. I think i's as good as `1203` and makes some code go away. WDYT?

Reply:
Yeah I like that. I'll see what I can do.

Reply:
Ah, but this does make `Dict` access a little more challenging. Now instead of doing `dict[1203]` you have to do `dict[[12, 3]]`.

Reply:
I'll pull it out into a constant for now.